charms/istio-pilot/src/charm.py

```diff
@@ -49,6 +41,10 @@ def __init__(self, *args):
     def install(self, event):
         """Install charm."""
 
+        if not isinstance(is_leader := self._check_is_leader(), ActiveStatus):
```

Comment:
Yeah I see your point on readability, but I don't know if booleans are a net improvement.  And I understand (although not sure I fully agree with) **setting** status in the handlers rather than helpers, but I don't see how removing information from the return of the handler improves the situation.  If anything I think it just leads to duplicate code (if status setting only happens in handlers, but both `relation_changed` and `update_status` need to validate relations are still ok, they will have duplicate code).

`ops` provides `Status`es that convey the entire message we want (`pass` (`Active`) vs `problem` (`Error`, `Waiting`, ...), and a message as to why).  Returning only `True`/`False` feels too simplistic.  For example, since the caller of a helper can only tell if the helper passed or failed, helpers can only check for one problem if the calling code wants to surface a descriptive error.  Rich status returns also felt convenient for things like [validating interfaces](https://github.com/canonical/istio-operators/blob/ee30f5162364c226a19540fc6d13b00ec20c9d61/charms/istio-pilot/src/charm.py#L272).  

We already provide richer failures through `Exception`s in [SDI errors](https://github.com/canonical/serialized-data-interface/blob/b27f5ce52a0de2e6cd885a023bbce2b37a43611b/serialized_data_interface/__init__.py#L21), but in those cases we're less opinionated about what the error means as it needs to be generic.  To me returning `Status` objects just feels like a more opinionated version of that pattern and since the entire code is in a single application we can be opinionated.  

I'm also not sure going from:
```python
        if not isinstance(is_leader := self._check_is_leader(), ActiveStatus):
            self.model.unit.status = is_leader
            return
```

to

```python
        if not self._check_is_leader():  # (or just self.unit.is_leader())
            self.model.unit.status = WaitingStatus("Waiting for leadership")
            return
```

is a simplification.  If we want handlers that read with their intent but hide their implementation details, I think we take the helpers-handle-status further:

```python
  def _is_activestatus(self, status, set_status_if_not_active=False):
    if isinstance(status, ActiveStatus):
      return True
    else:
      if set_status_if_not_active:
        self.model.unit.status = status
      return False

  def install(self, event):
    ...
    if self._is_activestatus(self._check_is_leader(), set_status_if_not_active=True):
      return

    if self._is_activestatus(self._check_required_relations(), set_status_if_not_active=True):
      return
    ...

  def handle_relation(self, event):
      if self._is_activestatus(self._check_required_relations(), set_status_if_not_active=True):
        return
      ...
```

In this case the helper is setting the status, but it isn't like that is hidden.  And then we can encapsulate the mapping of `state->status` in helpers and share it across multiple handlers

Reply:
> I'm also not sure going from:
[...]
to
[...]
is a simplification. 

I actually do think that the second code block is a lot cleaner to read. Just reading the code out loud I'm going from "if not isinstance" to "if not self unit is leader" (we should definitely use `self.unit.is_leader()`. This feels a lot more intuitive to me. Additionally, if I  use the returned status I don't know what kind of status to expect while in the second block I know exactly that I will pass to `WaitingStatus`.

Reply:
> We already provide richer failures through Exceptions in SDI errors, but in those cases we're less opinionated about what the error means as it needs to be generic. To me returning Status objects just feels like a more opinionated version of that pattern and since the entire code is in a single application we can be opinionated

Let's keep in mind that these rich failures are one of the reasons we are currently in the situation that we are. I would be careful of extending that pattern in all our codebase.